src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+#[derive(Debug)]
+pub enum Error {
+    /// The insertion failed because the new device overlapped with an old device.
+    Overlap,
+    /// PIO request is none.
+    NonePIORequest,
+    /// The insertion failed because device already exists.
+    Exist,
+    /// The removing fails because the device doesn't exist.
+    NonExist,
+    /// Irq allocation failed.
+    IrqAllocate(AllocatorError),
+    /// Instance id allocation failed.
+    InstanceIdAllocate(AllocatorError),
+    /// Address allocation failed.
+    AddressAllocate(AllocatorError),
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System device manager serving for all devices management and VM exit handling.
+pub struct DeviceManager {
+    /// System allocator reference.
+    resource: Arc<Mutex<SystemAllocator>>,
+    /// Devices information mapped by instance id.
+    devices: HashMap<u32, DeviceDescriptor>,
+    /// Range mapping for VM exit mmio operations.
+    mmio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+}
+
+impl DeviceManager {
+    /// Create a new `DeviceManager` with a `SystemAllocator` reference which would be
+    /// used to allocate resource for devices.
+    pub fn new(resource: Arc<Mutex<SystemAllocator>>) -> Self {
+        DeviceManager {
+            resource,
+            devices: HashMap::new(),
+            mmio_bus: BTreeMap::new(),
+            pio_bus: BTreeMap::new(),
+        }
+    }
+
+    fn insert(&mut self, dev: DeviceDescriptor) -> Result<()> {
+        // Insert if the key is non-present, else report error.
+        if self.devices.get(&(dev.instance_id)).is_some() {
+            return Err(Error::Exist);
+        }
+        self.devices.insert(dev.instance_id, dev);
+        Ok(())
+    }
+
+    fn remove(&mut self, instance_id: u32) -> Option<DeviceDescriptor> {
+        self.devices.remove(&instance_id)
+    }
+
+    fn device_descriptor(
+        &self,
+        id: u32,
+        name: String,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: Vec<IoResource>,
+        irq: Option<IrqResource>,
+    ) -> DeviceDescriptor {
+        DeviceDescriptor::new(id, name, dev.clone(), parent_bus, resources, irq)
+    }
+
+    // Allocate IO and instance id resources.
+    // Return a Result wrapper of instance id.
+    fn allocate_resources(&mut self, resources: &mut Vec<IoResource>) -> Result<(u32)> {
+        let id = self
+            .resource
+            .lock()
+            .expect("failed to acquire lock.")
+            .allocate_instance_id()
+            .map_err(Error::InstanceIdAllocate)?;
+
+        for res in resources.iter_mut() {
+            match res.res_type {
+                IoType::Pio => {
+                    if res.addr.is_none() {
+                        return Err(Error::NonePIORequest);
```

Comment:
Allocated resources, such as the id, should be released on errors.

Reply:
Oh, right!